packages/vx-demo/src/sandboxes/vx-tooltip/Example.tsx

```diff
@@ -0,0 +1,244 @@
+};
+
+export default function Example({ width, height, showControls = true }: TooltipProps) {
+  // bounds of the container are needed to convert page coordinates to container coordinates
+  const [ref, ownBounds] = useMeasure({ scroll: true });
+  const [detectBounds, setDetectBounds] = useState(true);
+  const [renderInPortal, setRenderInPortal] = useState(false);
+
+  const {
+    showTooltip,
+    hideTooltip,
+    tooltipOpen,
+    tooltipData,
+    tooltipLeft,
+    tooltipTop,
+  } = useTooltip<TooltipData>({
+    // initial tooltip state
+    tooltipOpen: true,
+    tooltipLeft: width / 3,
+    tooltipTop: height / 3,
+    tooltipData: {
+      text: 'Move me with your mouse or finger',
+      containerX: width / 3,
+      containerY: height / 3,
+      pageX: 0,
+      pageY: 0,
+    },
+  });
+
+  // event handlers
+  const handleMouseMove = useCallback(
+    (event: React.MouseEvent | React.TouchEvent) => {
+      const pageX = 'pageX' in event ? event.pageX : 0;
+      const pageY = 'pageY' in event ? event.pageY : 0;
+      const containerX = ('clientX' in event ? event.clientX : 0) - ownBounds.left;
+      const containerY = ('clientY' in event ? event.clientY : 0) - ownBounds.top;
+
+      showTooltip({
+        tooltipLeft: renderInPortal ? pageX : containerX,
+        tooltipTop: renderInPortal ? pageY : containerY,
+        tooltipData: {
+          containerX,
+          containerY,
+          pageX,
+          pageY,
+          text: detectBounds
+            ? 'I detect my container boundary'
+            : 'I will get clipped by my container',
+        },
+      });
+    },
+    [showTooltip, ownBounds, detectBounds, renderInPortal],
+  );
+
+  const TooltipComponent = detectBounds ? TooltipWithBounds : Tooltip;
+  const TooltipWrapper = renderInPortal ? Portal : React.Fragment;
+
+  return (
+    <>
+      <div
+        ref={ref}
+        className="tooltip-example"
+        style={{ width, height }}
+        onMouseMove={handleMouseMove}
+        onTouchMove={handleMouseMove}
+      >
+        {tooltipOpen ? (
+          <>
+            <div
+              className="position-indicator"
+              style={{
+                width: positionIndicatorSize,
+                height: positionIndicatorSize,
+                transform: `translate(${(tooltipData?.containerX ?? 0) -
+                  positionIndicatorSize / 2}px, ${(tooltipData?.containerY ?? 0) -
+                  positionIndicatorSize / 2}px)`,
+              }}
+            />
+            <div
+              className="crosshair horizontal"
+              style={{ transform: `translateY(${tooltipData?.containerY ?? 0}px)` }}
+            />
+            <div
+              className="crosshair vertical"
+              style={{ transform: `translateX(${tooltipData?.containerX ?? 0}px)` }}
+            />
+            <TooltipWrapper>
+              <TooltipComponent
+                key={Math.random()} // needed for bounds to update correctly
```

Comment:
> Keys should be stable, predictable, and unique. Unstable keys (like those produced by Math.random()) will cause many component instances and DOM nodes to be unnecessarily recreated, which can cause performance degradation and lost state in child components.
https://reactjs.org/docs/reconciliation.html#trade-offs

Does it have to be recreated all the time?

Reply:
yeah I think unfortunately for now, we do this in most of our demos. This forces a re-mount (😱) to force-update the bounds the component has, which basically change on every mouse move. Without this, if you start the tooltip on the left of the chart and move it to the right edge, it still thinks it has room because it's using the initial bounds. 

This really can't be that great for perf because of the remount and because [calling `.getBoundingClientRect()` forces layout/reflow](https://gist.github.com/paulirish/5d52fb081b3570c81e3a#getting-box-metrics). I tried to switch to `use-measure` for the boundary detection, but it only uses `scroll` listeners + `ResizeObserver`, so it doesn't re-compute on mouse move.

Reply:
left [a comment on `react-use-measure`](https://github.com/react-spring/react-use-measure/issues/14#issuecomment-649138515) about it cc @techniq

Reply:
`IntersectionObserver` seems promising.  If `react-use-measure` doesn't add something, we might be able to only update the `key` when the observer fires... to force an update.

I initially thought we could pre-compute the bounds based on the parent and tooltip size, but sinse the tooltip size can change size based on the contents, this wouldn't work.

Reply:
> I initially thought we could pre-compute the bounds based on the parent and tooltip size, but since the tooltip size can change size based on the contents, this wouldn't work.

ah this is interesting, currently we use the `bounds.left/top` attributes of the tooltip, rather than `bounds.width/height` (which should be more stable). I took a shot at updating to use the latter and not updating `key`, but hit issues in the portal where `width/height` of the parent can be `0` inside a portal 😭